app/javascript/components/ProductEdit/useAutosaveProduct.ts

```diff
@@ -0,0 +1,63 @@
+import * as React from "react";
+import isEqual from "lodash/isEqual";
+import { Product } from "./state";
+
+type Options = {
+  product: Product;
+  lastSavedProductRef: React.MutableRefObject<Product>;
+  save: () => Promise<void>;
+  saving: boolean;
+  isBlocked: boolean;
+};
+
+export const useAutosaveProduct = ({
+  product,
+  lastSavedProductRef,
+  save,
+  saving,
+  isBlocked,
+}: Options) => {
+  const debounceTimer = React.useRef<number | null>(null);
+  const intervalTimer = React.useRef<number | null>(null);
+
+  // TODO(Tri): Optimize with a dirtyRef instead of a deep-equality check
+  const isDirty = !isEqual(product, lastSavedProductRef.current);
+
+  // TODO(Tri): Handle `const autosaveErrorCount = React.useRef(0);`
+
+  const clearDebounce = () => {
+    if (debounceTimer.current !== null) {
+      clearTimeout(debounceTimer.current);
+      debounceTimer.current = null;
+    }
+  };
+
+  const triggerSave = React.useCallback(() => {
```

Comment:
Centralized save trigger with guards to avoid saving during **uploads**,
**in-progress saves**, or when **no changes** exist.